README.md
# OpenCensus Zenoss Exporter for Python

A Python library intended to be used in Python applications instrumented with
OpenCensus to export stats to Zenoss.

## Status

This library is in an alpha stage, and the API is subject to change.

## Installation

The package can be installed with the following command.

```shell script
pip install opencensus-ext-zenoss
```

## Usage

The following example shows how to configure the exporter.

```python
from opencensus.ext.zenoss import stats_exporter as zenoss
from opencensus.stats import stats as stats_module

stats = stats_module.stats
view_manager = stats.view_manager

exporter = zenoss.new_stats_exporter(
    zenoss.Options(api_key="YOUR-ZENOSS-API-KEY"))

view_manager.register_exporter(exporter)
```

## Options

The following options are available when creating a stats exporter.

* `address`: Zenoss API address. Default is https://api.zenoss.io.
* `api_key`: Zenoss API key.
* `source`: Added as a tag to all sent metrics. Recommended.
* `extra_tags`: Map of additional tags to add to all sent metrics. Default is {}.
* `insecure_tls`: Set to True to disable server certification verification. Default is False.

Example Application
-------------------

The following is a complete example of an application that will write to a
measure once per second for 100 seconds. A distribution view is created for
measure that will be exported to Zenoss once every 10 seconds.

```python
import os
import random
import sys
import time

from opencensus.ext.zenoss import stats_exporter as zenoss
from opencensus.stats import aggregation as aggregation_module
from opencensus.stats import measure as measure_module
from opencensus.stats import stats as stats_module
from opencensus.stats import view as view_module
from opencensus.tags import tag_map as tag_map_module

# Setup aliases to make working with OpenCensus easier.
stats = stats_module.stats
view_manager = stats.view_manager
stats_recorder = stats.stats_recorder

# Create a measure.
m_latency_ms = measure_module.MeasureFloat(
    "task_latency", "The task latency in milliseconds", "ms")

# Create a view using the measure.
latency_view = view_module.View(
    "task_latency_distribution",
    "The distribution of the task latencies",
    [],
    m_latency_ms,
    # Latency in buckets: [>=0ms, >=100ms, >=200ms, >=400ms, >=1s, >=2s, >=4s]
    aggregation_module.DistributionAggregation(
        [100.0, 200.0, 400.0, 1000.0, 2000.0, 4000.0]))


def main():
    address = os.environ.get("ZENOSS_ADDRESS", zenoss.DEFAULT_ADDRESS)
    api_key = os.environ.get("ZENOSS_API_KEY")
    if not api_key:
        sys.exit("ZENOSS_API_KEY must be set")

    # Create Zenoss exporter.
    exporter = zenoss.new_stats_exporter(
        options=zenoss.Options(
            address=address,
            api_key=api_key,
            source="app.example.com"),
        interval=10)

    # Register Zenoss exporter.
    view_manager.register_exporter(exporter)

    # Register our example view.
    view_manager.register_view(latency_view)

    # Prepare measurement map, and tag map we can reuse for each sample.
    measurement_map = stats_recorder.new_measurement_map()
    tag_map = tag_map_module.TagMap()

    # Record one random measurement each second for 100 seconds.
    print("Recording measurements:")
    for i in range(100):
        ms = random.random() * 5 * 1000
        print("  - latency {}:{}".format(i, ms))
        measurement_map.measure_float_put(m_latency_ms, ms)
        measurement_map.record(tag_map)
        time.sleep(1)


if __name__ == "__main__":
    import logging

    # Setup default logging configuration.
    logging.basicConfig()

    # Change level of Zenoss exporter's logging to debug.
    logging.getLogger("opencensus.ext.zenoss").setLevel(logging.DEBUG)

    main()
```

## Useful Links

* For more information about Zenoss, visit: <https://zenoss.com>.
* For more information about OpenCensus, visit: <https://opencensus.io>.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## v0.0.1
- Initial alpha release.


opencensus/ext/zenoss/stats_exporter.py
# Copyright 2020, Zenoss, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import datetime
import json
import logging
import requests
import time
import urllib3

from opencensus.metrics import transport
from opencensus.metrics.export import metric as metric_module
from opencensus.metrics.export import metric_descriptor
from opencensus.stats import stats

DEFAULT_ADDRESS = "https://api.zenoss.io"
DEFAULT_INTERVAL = 60
DEFAULT_SOURCE_TYPE = "zenoss/opencensus-python-exporter"
API_KEY_FIELD = "zenoss-api-key"
SOURCE_FIELD = "source"
SOURCE_TYPE_FIELD = "source-type"
DESCRIPTION_FIELD = "description"
UNITS_FIELD = "units"

EPOCH = datetime.datetime(1970, 1, 1, tzinfo=datetime.timezone.utc)


class Options(object):
    """Exporter configuration options."""

    def __init__(
            self,
            address=DEFAULT_ADDRESS,
            api_key=None,
            source=None,
            extra_tags=None,
            insecure_tls=False,
    ):
        self.address = address
        self.api_key = api_key
        self.source = source
        self.extra_tags = extra_tags or {}
        self.insecure_tls = insecure_tls

        if SOURCE_TYPE_FIELD not in self.extra_tags:
            self.extra_tags[SOURCE_TYPE_FIELD] = DEFAULT_SOURCE_TYPE

        if source and SOURCE_FIELD not in self.extra_tags:
            self.extra_tags[SOURCE_FIELD] = source

        if self.insecure_tls:
            urllib3.disable_warnings()


class ZenossStatsExporter(object):
    """Stats exporter for Zenoss."""

    def __init__(self, options=None):
        self.options = options or Options
        self.logger = logging.getLogger(__name__)

    def export_metrics(self, metrics):
        now = datetime.datetime.utcnow()
        tagged_metrics = []

        def add_metric(name, timestamp, value, tags):
            tagged_metrics.append({
                "metric": name,
                "timestamp": int(datetime_timestamp(timestamp or now) * 1000),
                "value": value,
                "tags": tags,
            })

        for metric in metrics:
            if not isinstance(metric, metric_module.Metric):
                continue

            descriptor = metric.descriptor

            for ts in metric.time_series:
                tags = {}
                for k, v in zip(descriptor.label_keys, ts.label_values):
                    tags[k.key] = v.value

                if descriptor.description:
                    tags[DESCRIPTION_FIELD] = descriptor.description

                if descriptor.unit:
                    tags[UNITS_FIELD] = descriptor.unit

                # extra_tags override metric tags.
                for k, v in self.options.extra_tags.items():
                    tags[k] = v

                for point in ts.points:
                    value = point.value

                    # "Regular" metric types.
                    if descriptor.type in (
                            metric_descriptor.MetricDescriptorType.CUMULATIVE_DOUBLE,
                            metric_descriptor.MetricDescriptorType.GAUGE_DOUBLE,
                            metric_descriptor.MetricDescriptorType.CUMULATIVE_INT64,
                            metric_descriptor.MetricDescriptorType.GAUGE_INT64):
                        add_metric(descriptor.name, point.timestamp, value.value, tags)

                    # Distribution metric types.
                    elif descriptor.type in (
                            metric_descriptor.MetricDescriptorType.CUMULATIVE_DISTRIBUTION,
                            metric_descriptor.MetricDescriptorType.GAUGE_DISTRIBUTION):
                        for s, v in {
                            "count": value.count,
                            "sum": value.sum,
                            "ss": value.sum_of_squared_deviation,
                            "mean": 0 if not value.count else value.sum / value.count,
                        }.items():
                            add_metric(
                                "{}/{}".format(descriptor.name, s),
                                point.timestamp, v, tags)

                    # Summary metric type.
                    elif descriptor.type == metric_descriptor.MetricDescriptorType.SUMMARY:
                        for s, v in {
                            "count": point.count,
                            "sum": point.sum,
                        }.items():
                            add_metric(
                                "{}/{}".format(descriptor.name, s),
                                point.timestamp, v, tags)

                    # Currently no other types, but log to be safe.
                    else:
                        self.logger.warning("unknown metric type: %s", metric.descriptor.type)

        if len(tagged_metrics) > 0:
            self.send_tagged_metrics(tagged_metrics)

    def send_tagged_metrics(self, tagged_metrics):
        metrics_url = "{}/v1/data-receiver/metrics".format(self.options.address)
        headers = {}
        if self.options.api_key:
            headers[API_KEY_FIELD] = self.options.api_key

        try:
            r = requests.post(
                metrics_url,
                headers=headers,
                verify=not self.options.insecure_tls,
                data=json.dumps({"taggedMetrics": tagged_metrics}))

            if not r.ok:
                r.raise_for_status()

            self.logger.debug("sent metrics: %s", r.json())
        except Exception as e:
            self.logger.error("failed to send metrics: %s", e)


def new_stats_exporter(options=None, interval=DEFAULT_INTERVAL):
    """Return a stats exporter and running transport thread."""
    exporter = ZenossStatsExporter(options=options)
    transport.get_exporter_thread([stats.stats], exporter, interval=interval)
    return exporter


def datetime_timestamp(dt):
    """Return POSIX timestamp as float.

    This is essentially Python 3.3's datetime.timestamp() function. I've
    moved it here for Python 2 compatibility.

    """
    if dt.tzinfo is None:
        return time.mktime(
            (dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second, -1, -1, -1)
        ) + dt.microsecond / 1e6
    else:
        return (dt - EPOCH).total_seconds()


examples/example.py
# Copyright 2020, Zenoss, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import random
import sys
import time

from opencensus.ext.zenoss import stats_exporter as zenoss
from opencensus.stats import aggregation as aggregation_module
from opencensus.stats import measure as measure_module
from opencensus.stats import stats as stats_module
from opencensus.stats import view as view_module
from opencensus.tags import tag_map as tag_map_module

# Setup aliases to make working with OpenCensus easier.
stats = stats_module.stats
view_manager = stats.view_manager
stats_recorder = stats.stats_recorder

# Create a measure.
m_latency_ms = measure_module.MeasureFloat(
    "task_latency", "The task latency in milliseconds", "ms")

# Create a view using the measure.
latency_view = view_module.View(
    "task_latency_distribution",
    "The distribution of the task latencies",
    [],
    m_latency_ms,
    # Latency in buckets: [>=0ms, >=100ms, >=200ms, >=400ms, >=1s, >=2s, >=4s]
    aggregation_module.DistributionAggregation(
        [100.0, 200.0, 400.0, 1000.0, 2000.0, 4000.0]))


def main():
    address = os.environ.get("ZENOSS_ADDRESS", zenoss.DEFAULT_ADDRESS)
    api_key = os.environ.get("ZENOSS_API_KEY")
    if not api_key:
        sys.exit("ZENOSS_API_KEY must be set")

    # Create Zenoss exporter.
    exporter = zenoss.new_stats_exporter(
        options=zenoss.Options(
            address=address,
            api_key=api_key,
            source="app.example.com"),
        interval=10)

    # Register Zenoss exporter.
    view_manager.register_exporter(exporter)

    # Register our example view.
    view_manager.register_view(latency_view)

    # Prepare measurement map, and tag map we can reuse for each sample.
    measurement_map = stats_recorder.new_measurement_map()
    tag_map = tag_map_module.TagMap()

    # Record one random measurement each second for 100 seconds.
    print("Recording measurements:")
    for i in range(100):
        ms = random.random() * 5 * 1000
        print("  - latency {}:{}".format(i, ms))
        measurement_map.measure_float_put(m_latency_ms, ms)
        measurement_map.record(tag_map)
        time.sleep(1)


if __name__ == "__main__":
    import logging

    # Setup default logging configuration.
    logging.basicConfig()

    # Change level of Zenoss exporter's logging to debug.
    logging.getLogger("opencensus.ext.zenoss").setLevel(logging.DEBUG)

    main()
